decisions/0001-file-sharing.md

```diff
@@ -0,0 +1,103 @@
+---
+# These are optional metadata elements. Feel free to remove any of them.
+status: "proposed"
+date: 2025-02-24
+decision-makers: tbd
+consulted: tbd
+informed: tbd
+---
+
+# Raw data / file sharing
+
+## Context and Problem Statement
+
+Loculus is currently only able to share consensus sequences but not the raw reads. However, such data are of high value. In case of clinical samples where consensus sequences are available, raw data are useful for deeper analysis, validation of results, and the detection of low-frequency variants. In the case of environmental samples, such as those from wastewater, consensus sequences cannot be derived and raw data are the only way to analyze the sample.
+
+The raw read data of a sample is significantly larger than a consensus sequence. For example, a consensus sequence of a SARS-CoV-2 sample is about 30 KB, while the raw data is around hundred MB to a few GB. We therefore need an approach to store and share large files.
+
+From the user perspective, depending on the instance, it should be possible to upload metadata, consensus sequences and/or raw data. The consensus sequences are, as now, stored in a FASTA. The raw data can be a FASTQ, a BAM file or something else – what exactly one has/can upload depends also on the instance. The maintainer-configurable pipeline will take the user data and return processed data: this could be only consensus sequences and metadata (i.e., some tabular data), or include some other unaligned raw data (e.g., the raw reads after dehumanizing), aligned raw reads, or other type of files. The user will be able to download all processed data. Aligned reads may be passed into SILO for fast querying.
```

Comment:
> a FASTQ, a BAM file or something else

will there always be one file per raw read sample or will there ever be multiple files?

Reply:
There will sometimes be multiple files

Reply:
Also good to already mention that ENA actually does quite some validation (or so it appears from the docs) on the metadata fields relating to raw reads - we will probably have to do this as well

Reply:
Yes, it can be multiple files. For the Loculus backend and the architecture that we discuss here, I don't think that we need to limit/define the set of files that can be accepted.

Reply:
It was one presigned URL per file, but that doesn't mean there can't be multiple

Reply:
The pre-signed URL approach also works with multiple files. Both the user and the pipeline can upload multiple files by requesting multiple pre-signed URLs. Hereby, I envision that you can specify a number of files that you want to upload when calling `/files/request-uploads` to get that number of URLs. Afterward, when calling `/submit`, `/revise` or `/submit-processed-data`, it is possible to provide the IDs of multiple files.

> I guess when we allow multiple files we must specify that users must upload them in one folder and they give us a link to the folder.

There is no concept of a folder in S3 and it is not possible to generate one pre-signed URL for a folder but only for a specific object.

Reply:
Ah ok that was were the requesting multiple URLs came in... I thought it was just if the user has multiple raw reads.

I got to admit I do think this becomes quite complex and it gets even harder for users to prepare their data in advance. It is possible to submit folders in aws - so I guess this is a limitation of signed URLs: 

```
aws s3 cp /local/folder/path s3://your-bucket-name/ --recursive
```

Reply:
But `aws s3 cp /local/folder/path s3://your-bucket-name/ --recursive` is not done in a single HTTP request, is it? I assume that it is split up into multiple requests.

I envision that most users won't need to care about the API but will just use a CLI or web client (that we will certainly implement in the future).

Reply:
true! The endpoint would have to take a zipped file (it is possible to unzip while streaming without loading the file into memory so we could still do some checks on the data even)